libbpfgo.go

```diff
@@ -475,491 +389,71 @@ func (m *Module) InitGlobalVariable(name string, value interface{}) error {
-		return nil, fmt.Errorf("failed to batch lookup and delete values %v in map %s: ret %d (err: %s)", keys, b.name, ret, errC)
-	}
-
-	// Either some or all entries were read and deleted.
-	parsedVals := collectBatchValues(values, processed, b.ValueSize())
-	return parsedVals, nil
-}
-
-func collectBatchValues(values []byte, count uint32, valueSize int) [][]byte {
-	var value []byte
-	var collected [][]byte
-	for i := 0; i < int(count*uint32(valueSize)); i += valueSize {
-		value = values[i : i+valueSize]
-		collected = append(collected, value)
-	}
-	return collected
-}
-
-// UpdateBatch updates multiple elements in the map by specified keys and their corresponding values.
-//
-// The first argument, keys, is a pointer to an array or slice of keys which will be updated using the second argument, values.
-// It returns the associated error if any occurred.
-//
-// The last argument, count, is the number of keys to update. Passing an argument that greater than the number of keys
-// in the map will cause the function to return a syscall.EPERM as an error.
-func (b *BPFMap) UpdateBatch(keys, values unsafe.Pointer, count uint32) error {
-	countC := C.uint(count)
+	if !m.loaded {
+		bpfMap.bpfMapLow = &BPFMapLow{
+			fd:   -1,
+			info: &BPFMapInfo{},
+		}
 
-	opts := BPFMapBatchOpts{
-		Sz:        uint64(unsafe.Sizeof(BPFMapBatchOpts{})),
-		ElemFlags: C.BPF_ANY,
-		Flags:     C.BPF_ANY,
+		return bpfMap, nil
 	}
 
-	errC := C.bpf_map_update_batch(b.fd, keys, values, &countC, bpfMapBatchOptsToC(&opts))
-	if errC != 0 {
-		sc := syscall.Errno(-errC)
-		if sc != syscall.EFAULT {
-			if uint32(countC) != count {
-				return fmt.Errorf("failed to update ALL elements in map %s, updated (%d/%d): %w", b.name, uint32(countC), count, sc)
-			}
+	fd := bpfMap.FileDescriptor()
+	info, err := GetMapInfoByFD(fd)
+	if err != nil {
+		// Compatibility Note: Some older kernels lack BTF (BPF Type Format)
+		// support for specific BPF map types. In such scenarios, libbpf may
+		// fail (EPERM) when attempting to retrieve information for these maps.
+		// Reference: https://elixir.bootlin.com/linux/v5.15.75/source/tools/lib/bpf/gen_loader.c#L401
+		//
+		// However, we can still get some map info from the BPF map high level API.
+		bpfMap.bpfMapLow = &BPFMapLow{
+			fd: fd,
+			info: &BPFMapInfo{
+				Type:                  bpfMap.Type(),
+				ID:                    0,
+				KeySize:               uint32(bpfMap.KeySize()),
+				ValueSize:             uint32(bpfMap.ValueSize()),
+				MaxEntries:            bpfMap.MaxEntries(),
+				MapFlags:              uint32(bpfMap.MapFlags()),
+				Name:                  bpfMap.Name(),
+				IfIndex:               bpfMap.IfIndex(),
+				BTFVmlinuxValueTypeID: 0,
+				NetnsDev:              0,
+				NetnsIno:              0,
+				BTFID:                 0,
+				BTFKeyTypeID:          0,
+				BTFValueTypeID:        0,
+				MapExtra:              bpfMap.MapExtra(),
+			},
```

Comment:
This error was detected while implementing the policies version on Tracee running kernel 5.15.75. It doesn't occur in 6.1.44: https://elixir.bootlin.com/linux/v6.1.44/source/tools/lib/bpf/gen_loader.c#L470